packages/hop-node/src/db/TransferRootsDb.ts

```diff
@@ -75,57 +73,124 @@ const invalidTransferRoots: Record<string, boolean> = {
+        if (!item?.transferRootId) {
+          resolve(null)
+          return
+        }
+        const { transferRootId } = item
+        await this.subDbIncompletes._update(transferRootId, { transferRootId })
+        await this.subDbIncompletes.deleteById(key)
+        resolve(null)
+      }))
+    }
+
+    await Promise.all(promises)
+
+    // re-index db keys by transfer root id
+    items = await this.getKeyValues()
+    promises = []
+    for (const { key, value } of items) {
+      promises.push(new Promise(async (resolve) => {
+        if (!value?.transferRootId || key === value?.transferRootId) {
+          resolve(null)
+          return
+        }
+        const { transferRootId } = value
+        await this._update(transferRootId, value)
+        await this.deleteById(key)
+        resolve(null)
+      }))
+    }
 
-    this.ready = true
-    this.logger.debug('db ready')
+    await Promise.all(promises)
   }
 
   async updateIncompleteItem (item: Partial<TransferRoot>) {
     if (!item) {
       this.logger.error('expected item', item)
       return
     }
-    const { transferRootHash } = item
-    if (!transferRootHash) {
-      this.logger.error('expected transferRootHash', item)
+    const { transferRootId } = item
+    if (!transferRootId) {
+      this.logger.error('expected transferRootId', item)
       return
     }
     const isIncomplete = this.isItemIncomplete(item)
-    const exists = await this.subDbIncompletes.getById(transferRootHash)
+    const exists = await this.subDbIncompletes.getById(transferRootId)
     const shouldUpsert = isIncomplete && !exists
     const shouldDelete = !isIncomplete && exists
     if (shouldUpsert) {
-      await this.subDbIncompletes._update(transferRootHash, { transferRootHash })
+      await this.subDbIncompletes._update(transferRootId, { transferRootId })
     } else if (shouldDelete) {
-      await this.subDbIncompletes.deleteById(transferRootHash)
+      await this.subDbIncompletes.deleteById(transferRootId)
     }
   }
 
-  async trackTimestampedKey (transferRoot: Partial<TransferRoot>) {
-    const data = await this.getTimestampedKeyValueForUpdate(transferRoot)
-    if (data != null) {
-      const key = data.key
-      const transferRootHash = data.value.transferRootHash
-      this.logger.debug(`storing timestamped key. key: ${key} transferRootHash: ${transferRootHash}`)
-      const value = { transferRootHash }
-      await this.subDb._update(key, value)
-    }
-  }
-
-  async trackTimestampedKeyByTransferRootHash (transferRootHash: string) {
-    const transferRoot = await this.getByTransferRootHash(transferRootHash)
-    return await this.trackTimestampedKey(transferRoot)
-  }
-
   getTimestampedKey (transferRoot: Partial<TransferRoot>) {
-    if (transferRoot.committedAt && transferRoot.transferRootHash) {
-      const key = `transferRoot:${transferRoot.committedAt}:${transferRoot.transferRootHash}`
+    if (transferRoot.committedAt && transferRoot.transferRootId) {
```

Comment:
(I think this should be addressed in another PR)

It is not guaranteed that a transferRoot has a `committedAt` entry at this point in its lifecycle. I think there is an edge case here where a bonder calls `bondTransferRoot` (maliciously, for example) on a non-existent `transferId`. In this case `transferRoot.committedAt` will be undefined. I believe we _should_ handle this somehow, but I'm not sure how and I think it might be worth taking care of after this PR.

Reply:
That's a great point. Maybe we track transferRoots with no commitedAt in separate subdb and then delete it from this subdb when the entry does have a commitedAt?

Reply:
Not a bad idea. Let's discuss on standup.